File: 2022/day15/php/solution.php
#!/usr/bin/env php
<?php

/**
 * Day 15: Beacon Exclusion Zone
 *
 * Parse sensor/beacon positions and calculate coverage using Manhattan distance.
 * Part 1: Count positions in row y=2000000 that cannot contain a beacon
 * Part 2: Find the one uncovered position in 0-4000000 range
 */

function parseSensors(string $text): array {
    $sensors = [];
    $pattern = '/Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)/';

    foreach (explode("\n", trim($text)) as $line) {
        if (preg_match($pattern, $line, $matches)) {
            $sx = (int)$matches[1];
            $sy = (int)$matches[2];
            $bx = (int)$matches[3];
            $by = (int)$matches[4];
            $dist = abs($sx - $bx) + abs($sy - $by);
            $sensors[] = [$sx, $sy, $bx, $by, $dist];
        }
    }

    return $sensors;
}

function mergeRanges(array $ranges): array {
    if (empty($ranges)) {
        return [];
    }

    usort($ranges, function($a, $b) {
        return $a[0] <=> $b[0];
    });

    $merged = [$ranges[0]];

    for ($i = 1; $i < count($ranges); $i++) {
        $start = $ranges[$i][0];
        $end = $ranges[$i][1];
        $lastIdx = count($merged) - 1;

        if ($start <= $merged[$lastIdx][1] + 1) {
            $merged[$lastIdx][1] = max($merged[$lastIdx][1], $end);
        } else {
            $merged[] = [$start, $end];
        }
    }

    return $merged;
}

function getCoverageAtRow(array $sensors, int $row): array {
    $ranges = [];

    foreach ($sensors as $sensor) {
        list($sx, $sy, $bx, $by, $dist) = $sensor;

        $rowDist = abs($sy - $row);
        if ($rowDist > $dist) {
            continue;
        }

        $xSpread = $dist - $rowDist;
        $ranges[] = [$sx - $xSpread, $sx + $xSpread];
    }

    return mergeRanges($ranges);
}

function part1(string $text): int {
    $sensors = parseSensors($text);
    $targetRow = 2000000;

    $ranges = getCoverageAtRow($sensors, $targetRow);

    $total = 0;
    foreach ($ranges as $range) {
        $total += $range[1] - $range[0] + 1;
    }

    $beaconsOnRow = [];
    foreach ($sensors as $sensor) {
        list($sx, $sy, $bx, $by, $dist) = $sensor;
        if ($by === $targetRow) {
            $beaconsOnRow[$bx] = true;
        }
    }

    return $total - count($beaconsOnRow);
}

function part2(string $text): int {
    $sensors = parseSensors($text);
    $maxCoord = 4000000;

    for ($row = 0; $row <= $maxCoord; $row++) {
        $ranges = getCoverageAtRow($sensors, $row);

        $clipped = [];
        foreach ($ranges as $range) {
            list($start, $end) = $range;
            if ($end < 0 || $start > $maxCoord) {
                continue;
            }
            $clipped[] = [max(0, $start), min($maxCoord, $end)];
        }

        $clipped = mergeRanges($clipped);

        if (count($clipped) === 1 && $clipped[0][0] === 0 && $clipped[0][1] === $maxCoord) {
            continue;
        }

        if (count($clipped) > 1) {
            $x = $clipped[0][1] + 1;
        } elseif ($clipped[0][0] > 0) {
            $x = 0;
        } else {
            $x = $clipped[0][1] + 1;
        }

        return $x * 4000000 + $row;
    }

    return 0;
}

$scriptDir = dirname(__FILE__);
$inputFile = $scriptDir . '/../input.txt';
$text = file_get_contents($inputFile);

echo 'Part 1: ' . part1($text) . "\n";
echo 'Part 2: ' . part2($text) . "\n";

File: 2022/day15/java/Solution.java
import java.io.*;
import java.util.*;
import java.util.regex.*;

public class Solution {

    static class Sensor {
        int sx, sy, bx, by, dist;

        Sensor(int sx, int sy, int bx, int by) {
            this.sx = sx;
            this.sy = sy;
            this.bx = bx;
            this.by = by;
            this.dist = Math.abs(sx - bx) + Math.abs(sy - by);
        }
    }

    static List<Sensor> parseSensors(String text) {
        List<Sensor> sensors = new ArrayList<>();
        Pattern pattern = Pattern.compile(
            "Sensor at x=(-?\\d+), y=(-?\\d+): closest beacon is at x=(-?\\d+), y=(-?\\d+)"
        );

        for (String line : text.trim().split("\n")) {
            Matcher matcher = pattern.matcher(line);
            if (matcher.find()) {
                int sx = Integer.parseInt(matcher.group(1));
                int sy = Integer.parseInt(matcher.group(2));
                int bx = Integer.parseInt(matcher.group(3));
                int by = Integer.parseInt(matcher.group(4));
                sensors.add(new Sensor(sx, sy, bx, by));
            }
        }

        return sensors;
    }

    static List<int[]> getCoverageAtRow(List<Sensor> sensors, int row) {
        List<int[]> ranges = new ArrayList<>();

        for (Sensor s : sensors) {
            int rowDist = Math.abs(s.sy - row);
            if (rowDist > s.dist) {
                continue;
            }

            int xSpread = s.dist - rowDist;
            ranges.add(new int[]{s.sx - xSpread, s.sx + xSpread});
        }

        return mergeRanges(ranges);
    }

    static List<int[]> mergeRanges(List<int[]> ranges) {
        if (ranges.isEmpty()) {
            return new ArrayList<>();
        }

        ranges.sort((a, b) -> a[0] != b[0] ? a[0] - b[0] : a[1] - b[1]);

        List<int[]> merged = new ArrayList<>();
        merged.add(ranges.get(0).clone());

        for (int i = 1; i < ranges.size(); i++) {
            int[] current = ranges.get(i);
            int[] last = merged.get(merged.size() - 1);

            if (current[0] <= last[1] + 1) {
                last[1] = Math.max(last[1], current[1]);
            } else {
                merged.add(current.clone());
            }
        }

        return merged;
    }

    static long part1(String text) {
        List<Sensor> sensors = parseSensors(text);
        int targetRow = 2000000;

        List<int[]> ranges = getCoverageAtRow(sensors, targetRow);

        long total = 0;
        for (int[] range : ranges) {
            total += range[1] - range[0] + 1;
        }

        Set<Integer> beaconsOnRow = new HashSet<>();
        for (Sensor s : sensors) {
            if (s.by == targetRow) {
                beaconsOnRow.add(s.bx);
            }
        }

        return total - beaconsOnRow.size();
    }

    static long part2(String text) {
        List<Sensor> sensors = parseSensors(text);
        int maxCoord = 4000000;

        for (int row = 0; row <= maxCoord; row++) {
            List<int[]> ranges = getCoverageAtRow(sensors, row);

            List<int[]> clipped = new ArrayList<>();
            for (int[] range : ranges) {
                if (range[1] < 0 || range[0] > maxCoord) {
                    continue;
                }
                clipped.add(new int[]{Math.max(0, range[0]), Math.min(maxCoord, range[1])});
            }

            clipped = mergeRanges(clipped);

            if (clipped.size() == 1 && clipped.get(0)[0] == 0 && clipped.get(0)[1] == maxCoord) {
                continue;
            }

            long x;
            if (clipped.size() > 1) {
                x = clipped.get(0)[1] + 1;
            } else if (clipped.get(0)[0] > 0) {
                x = 0;
            } else {
                x = clipped.get(0)[1] + 1;
            }

            return x * 4000000L + row;
        }

        return -1;
    }

    public static void main(String[] args) throws IOException {
        String inputFile = "../input.txt";
        if (args.length > 0) {
            inputFile = args[0];
        }

        String text = new String(java.nio.file.Files.readAllBytes(
            java.nio.file.Paths.get(inputFile)));

        System.out.println("Part 1: " + part1(text));
        System.out.println("Part 2: " + part2(text));
    }
}
